### PR TITLE
Updates pubkey from bytes32 to bytes

### PR DESCRIPTION
**Pull request by IvanTheGreatDev**
Resolves issue #562 

---

# Changes pubkey from bytes32 to bytes 
The changes made modify all the `bytes32` to `bytes` and make sure its checks that it is a 48length byte array.
Also changes pubKey mapping to a hashPubKey mapping due to the changes required.

### CODE REVIEW

File: contracts/validator-registration-contract/validator_registration.go [hunk 1]
```diff
@@ -231,21 +231,21 @@ func (_ValidatorRegistration *ValidatorRegistrationCallerSession) UsedPubkey(arg
 // Deposit is a paid mutator transaction binding the contract method 0x881d2135.
 //
 // Solidity: function deposit(_pubkey bytes32, _withdrawalShardID uint256, _withdrawalAddressbytes32 address, _randaoCommitment bytes32) returns()
-func (_ValidatorRegistration *ValidatorRegistrationTransactor) Deposit(opts *bind.TransactOpts, _pubkey [32]byte, _withdrawalShardID *big.Int, _withdrawalAddressbytes32 common.Address, _randaoCommitment [32]byte) (*types.Transaction, error) {
```
Comment (terenc3t) [inline contracts/validator-registration-contract/validator_registration.go]: Did you regenerate `validator_registration.go` or manually changed it `[48]byte`?

Weird it still says `function deposit(_pubkey bytes32,...` at the top

**Comment by IvanTheGreatDev** [comment]
Changing the mapping from `bytes32` to `bytes` is actually not possible in Solidity since `bytes` is a dynamic type.

**Comment by prestonvanloon** [comment]
What is wrong with a dynamic type other than being wasteful? 

https://solidity.readthedocs.io/en/v0.4.24/types.html#dynamically-sized-byte-array

Can you store the pubkey as two byte32s and concat them?

**Comment by IvanTheGreatDev** [comment]
https://solidity.readthedocs.io/en/v0.4.25/types.html#mappings
Dynamic types can't be used as the key for mappings.
`Mapping types are declared as mapping(_KeyType => _ValueType). Here _KeyType can be almost any type except for a mapping, a dynamically sized array, a contract, an enum and a struct. _ValueType can actually be any type, including mappings.`

What would I concat them into though? I can't do a string (because dynamic length) or a number (48 bytes > 32 bytes).

**Comment by nisdas** [comment]
@IvanTheGreatDev Whats the status on this ?

**Comment by IvanTheGreatDev** [comment]
Still the same status. Solidity doesn't allow bytes as a key type for
mappings, so the max amount of bytes we can use is 32.

**Comment by nisdas** [comment]
@terenc3t @prestonvanloon What do you guys think ? Should we just store the hash of the pubkey instead in the mapping ?

**Comment by IvanTheGreatDev** [comment]
Alright, I changed it to just use the hash of the bytes array the user passes in. Good for review now.

**Pull request merged**